### ISSUE TITLE
Python3 support

### ISSUE THREAD

**Issue by Stvad** [robinhilliard/pipes#1]
Hey, thank you for creating this, just what I'm looking for 😃 

It'd be great if you could add Python3 support!
Currently, if I try to use it in Python3, I'm getting the following error:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-28ed7b59ff38> in <module>()
      5     return a * b
      6 
----> 7 @pipes
      8 def calc():
      9     print(1 >> add(2) >> times(3))  # prints 9

/usr/local/anaconda3/lib/python3.6/site-packages/pipeop/__init__.py in pipes(func)
     17 def pipes(func):
     18     # name of our replacement function
---> 19     pipe_func_name = '__pipes_{}'.format(func.func_code.co_name)
     20 
     21     # variable context where decorator added

AttributeError: 'function' object has no attribute 'func_code'
```

**Comment by robinhilliard** [comment]
Hi Stvad,

I'm not a regular Python 3 user but I wrote some tests and got them passing in Python 3.6.5 with a few minor tweaks. I've committed them to master if you want to try them out, if they work for you let me know and I'll update pypi.

Cheers,
Robin

**Comment by Stvad** [comment]
Hey @robinhilliard thank you for quick response!  
`I'm not a regular Python 3` you should migrate, it's nice here ;) 

The original example is indeed working now 🎆. I haven't played with the library in depth yet, but a first few comments on other things:
1) For the use cases of functions like `map` or `reduce` it seems it'd be better to pass the piped parameter as a last attribute and not the first one (can that be made configurable?).
2) When you have a function with only 1 parameter it's a bit confusing that you have to do the actual cal vs providing the function name (i.e. `range(5) >> sum()` vs  `range(5) >> sum`

**Comment by robinhilliard** [comment]
Hi Stvad,

Done, again on master – check it out. Given I'd already done terrible things to `>>` I thought I may as well do the same with `<<`, which now appends arguments as requested. I've also supported dropping braces from single argument functions. 

We've thought about Python 3 at work for a while but we have a lot of libraries and C dependencies to bring across and while I'm sure Python 3 is very schmick, 2.7 is a pretty comfortable neighbourhood :-).

Let me know when you've tried it out, if good I'll push it to Pypi. I've given you a shoutout at the bottom of the README, hope that's ok.

Cheers,
Robin

**Comment by Stvad** [comment]
Hey Robin, great, thank you for implementing this! :)

The chaining does not seem to work reliably for me:
`2>>print` works
`2>>pow(2)` works
`2>>pow(2)>>pow(3)` works
`2>>pow(2)>>print` does not
```
Traceback (most recent call last):
  File "test.py", line 13, in <module>
    test()
  File "/private/tmp/pipes/pipeop/__init__.py", line 10, in __pipes_test
    if isinstance(node.op, (LShift, RShift)):
TypeError: pow expected at least 2 arguments, got 1
```
Than you for the shoutout 🙂

Also, it's a bit awkward to test this in interactive shell (as I need to redefine a function with a new name each time, to make @pipes decoration process it a new), so I have another feature request to do the deduplication check in a different way if its possible 🙂(or to allow to pass a parameter to decorator to disable dedup check to allow for interactive usage use-case)

**Comment by robinhilliard** [comment]
Hi Stavd,

The chaining problem is caused by `print` being a statement, not a function, when used without braces. Another function would work.

I think I'll look into passing a `force_recompile` or similar argument to the decorator, because it's an edge case and I wouldn't want to try to detect the interactive shell and opaquely change the behaviour in response.

**Comment by Stvad** [comment]
Hey,

It is actually a function in python3 ;) and also the example #1 works (`2>>print`). 
But it's somehow related to the parenthesis thing (and to the type of the callable?):
`range(5)<<map(abs)>>list()>>print()` works for example
`range(5)<<map(abs)>>list` doesn't

**Comment by robinhilliard** [comment]
Chaining and `@pipes(False)` to prevent reading the cache should all be fixed on master, we're getting close :-).

**Comment by Stvad** [comment]
Yes we are :) I was able to break it again though 😛

`5<<lambda a: a**2<<print`
breaks with 

```
 /tmp/pipes   master  python3 test.py
  File "test.py", line 19
    5<<lambda a: a**2<<print
            ^
SyntaxError: invalid syntax
```

`5<<(lambda a: a**2)<<print` breaks with:

```
 ✘  /tmp/pipes   master  python3 test.py
Traceback (most recent call last):
  File "test.py", line 5, in <module>
    @pipes
  File "/private/tmp/pipes/pipeop/__init__.py", line 82, in pipes
    return pipes_decorator(func_or_cache_flag)
  File "/private/tmp/pipes/pipeop/__init__.py", line 64, in pipes_decorator
    tree = _PipeTransformer().visit(tree)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 308, in generic_visit
    value = self.visit(value)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 308, in generic_visit
    value = self.visit(value)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 317, in generic_visit
    new_node = self.visit(old_value)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/private/tmp/pipes/pipeop/__init__.py", line 20, in visit_BinOp
    col_offset=node.right.col_offset
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 308, in generic_visit
    value = self.visit(value)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/private/tmp/pipes/pipeop/__init__.py", line 24, in visit_BinOp
    node.right.args.insert(
AttributeError: 'arguments' object has no attribute 'insert'
```
and finally `5<<(lambda a: a**2)()<<print` works :)

**Comment by robinhilliard** [comment]
Just confirming you'd want to be able to drop the brackets on a single argument lambda,like we did with functions? Lambdas themselves would need to be bracketed for the parser to work on the initial parse, so we could do

`5 << (lambda a: a**2) << print`

but not

`5 << lambda a: a**2 << print`

**Comment by Stvad** [comment]
That's correct if the `5 << lambda a: a**2 << print` is not feasible, the `5 << (lambda a: a**2) << print` is definately nicer then `5 << (lambda a: a**2)() << print`

**Comment by robinhilliard** [comment]
Cool I'll look at it tonight (Sydney time).

**Comment by robinhilliard** [comment]
Done, tested, checked in on master and the readmes are updated. Let me know how you go.

**Comment by Stvad** [comment]
Looks good to me 👍

**Comment by Stvad** [comment]
ah, actually I forgot about the class use case 😛. It's the braces thing again, I'm afraid:

```python
from pipeop import pipes

class tc:
    def tp(self, arg):
        print(arg)

    @pipes
    def tf(self):
        range(-2,2)<<map(abs)<<sum<<self.tp


if __name__ == '__main__':
    tc().tf()
```

breaks with:

```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    class tc:
  File "test.py", line 7, in tc
    @pipes
  File "/private/tmp/pipes/pipeop/__init__.py", line 82, in pipes
    return pipes_decorator(func_or_cache_flag)
  File "/private/tmp/pipes/pipeop/__init__.py", line 64, in pipes_decorator
    tree = _PipeTransformer().visit(tree)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 308, in generic_visit
    value = self.visit(value)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 308, in generic_visit
    value = self.visit(value)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 317, in generic_visit
    new_node = self.visit(old_value)
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/ast.py", line 253, in visit
    return visitor(node)
  File "/private/tmp/pipes/pipeop/__init__.py", line 24, in visit_BinOp
    node.right.args.insert(
AttributeError: 'Attribute' object has no attribute 'args'
```
The version with braces works though (`range(-2,2)<<map(abs)<<sum<<self.tp()`)

**Comment by robinhilliard** [comment]
Spent some time getting my other ongoing project [eliXPlane](https://github.com/robinhilliard/elixplane) ready for a user group talk — back onto this now.

**Comment by robinhilliard** [comment]
Fixed in 0.2.1 - generalised handling of anything that isn't a function call on the right side of a pipe, so should deal with future, similar cases. I'll close this ticket and continue on #2.

**Comment by Stvad** [comment]
Cool! Hope the talk went well ;)
